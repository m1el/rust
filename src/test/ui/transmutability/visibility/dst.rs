//! The destination type, its fields, and its fields' types (and so on) must be visible from
//! `Context`.

#![feature(transmutability)]
#![allow(dead_code)]

// the field of `Dst` is private
mod dst_field_private {
    mod assert {
        use std::mem::BikeshedIntrinsicFrom;

        pub fn is_transmutable<Src, Dst, Context>()
        where
            Dst: BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
        {}
    }

    mod src {
        #[repr(C)] pub(in super) struct Zst;

        #[repr(C)] pub(in super) struct Src {
            pub(in super) field: Zst,
        }
    }

    mod dst {
        #[repr(C)] pub(in super) struct Zst;

        #[repr(C)] pub(in super) struct Dst {
            pub(self) field: Zst, // <- private field
        }
    }

    const _: () = {|| {
        struct Context;
        assert::is_transmutable::<src::Src, dst::Dst, Context>(); //~ ERROR not satisfied
    };};
}

// the type of `Dst` is private
mod dst_type_private {
    mod src {
        #[repr(C)] pub(in super) struct Zst;

        #[repr(C)] pub(in super) struct Src {
            pub(in super) field: Zst,
        }
    }

    mod dst {
        #[repr(C)] pub(in super) struct Zst;

        #[repr(C)] pub(self) struct Dst { // <- private type
            pub(in super) field: Zst,
        }

        use std::mem::BikeshedIntrinsicFrom;

        pub trait IsTransmutable<Src, Context> {}

        impl<Src, Context> IsTransmutable<Src, Context> for Dst
        where
            Dst: BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
        {}

        pub fn is_transmutable<Src, Context>()
        where
            Dst: IsTransmutable<Src, Context>
        {}
    }

    const _: () = {|| {
        pub(self) struct Context;
        dst::is_transmutable::<src::Src, Context>(); //~ ERROR not satisfied
    };};
}

// the type of `Dst`'s field is private
mod dst_field_type_private {
    mod src {
        #[repr(C)] pub(in super) struct Zst;

        #[repr(C)] pub(in super) struct Src {
            pub(in super) field: Zst,
        }
    }

    mod dst {
        #[repr(C)] pub(self) struct Zst; // <- private type

        #[repr(C)] pub(in super) struct Dst {
            pub(in super) field: Zst,
        }

        use std::mem::BikeshedIntrinsicFrom;

        pub trait IsTransmutable<Src, Context> {}

        impl<Src, Context> IsTransmutable<Src, Context> for Dst
        where
            Dst: BikeshedIntrinsicFrom<Src, Context, false, false, false, false>
        {}

        pub fn is_transmutable<Src, Context>()
        where
            Dst: IsTransmutable<Src, Context>
        {}
    }

    const _: () = {|| {
        pub(self) struct Context;
        dst::is_transmutable::<src::Src, Context>(); //~ ERROR not satisfied
    };};
}

fn main() {}
