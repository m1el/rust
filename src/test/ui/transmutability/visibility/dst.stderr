error[E0446]: private type `dst_type_private::dst::Dst` in public interface
  --> $DIR/dst.rs:66:9
   |
LL |           #[repr(C)] pub(self) struct Dst { // <- private type
   |                      -------------------- `dst_type_private::dst::Dst` declared as private
...
LL | /         pub fn is_transmutable<Src, Context>()
LL | |         where
LL | |             Dst: IsTransmutable<Src, Context>
   | |_____________________________________________^ can't leak private type

error[E0446]: private type `dst_field_type_private::dst::Zst` in public interface
  --> $DIR/dst.rs:92:13
   |
LL |         #[repr(C)] pub(self) struct Zst; // <- private type
   |                    --------------------- `dst_field_type_private::dst::Zst` declared as private
...
LL |             pub(in super) field: Zst,
   |             ^^^^^^^^^^^^^^^^^^^^^^^^ can't leak private type

error[E0446]: restricted type `dst_field_type_private::dst::Dst` in public interface
  --> $DIR/dst.rs:104:9
   |
LL |           #[repr(C)] pub(in super) struct Dst {
   |                      ------------------------ `dst_field_type_private::dst::Dst` declared as restricted
...
LL | /         pub fn is_transmutable<Src, Context>()
LL | |         where
LL | |             Dst: IsTransmutable<Src, Context>
   | |_____________________________________________^ can't leak restricted type

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0446`.
